lib/label.ml

```diff
@@ -0,0 +1,195 @@
+  match value with
+  | Some (op, v) -> f op v
+  | None -> Error (`Msg (Format.sprintf "Label `%s` requires a value." label))
+
+let requires_eq ~label ~op ~value f =
+  match op with
+  | Relation.Eq -> Ok (f value)
+  | _ ->
+    let msg =
+      Format.sprintf
+        "Label `%s` requires assignment using the `=` operator." label
+    in
+    Error (`Msg msg)
+
+let requires_eq_value ~label ~value f =
+  requires_value ~label ~value (fun op value ->
+      requires_eq ~label ~op ~value f)
+
+let interpret label value =
+  match label with
+  (* flags: labels without value *)
+  | "skip" ->
+    doesnt_accept_value ~label ~value Skip
+  | v when is_prefix ~prefix:"unset-" v ->
+    doesnt_accept_value ~label ~value (Unset (split_prefix ~prefix:"unset-" v))
+  (* labels requiring a value and any comparison operator *)
+  | "version" -> (
+      requires_value ~label ~value (fun op v ->
+          match Ocaml_version.of_string v with
+          | Ok v -> Ok (Version (op, v))
+          | Error (`Msg e) ->
+            let msg = Format.sprintf "Invalid `version` label value: %s." e in
+            Error (`Msg msg) ) )
+  (* non-deterministic accepts any value *)
+  | "non-deterministic" -> (
+      match value with
+      | Some (Relation.Eq, "output") -> Ok (Non_det `Output)
+      | Some (Relation.Eq, "command") -> Ok (Non_det `Command)
+      | Some (Relation.Eq, v) ->
+        let msg =
+          Format.sprintf
+            "%S is not a valid value for label `%s`. Valid values are <none>,\
+            \ %S and %S."
+            v label "command" "output"
+        in
+        Error (`Msg msg)
+      | Some _ ->
+        let msg =
+          Format.sprintf
+            "label `%s` requires assignment using the `=` operator." label
+        in
+        Error (`Msg msg)
+      | None -> Ok (Non_det `Output) )
+  (* labels requiring a value and '=' operator *)
+  | "dir" -> requires_eq_value ~label ~value (fun x -> Dir x)
+  | "source-tree" -> requires_eq_value ~label ~value (fun x -> Source_tree x)
+  | "file" -> requires_eq_value ~label ~value (fun x -> File x)
+  | "part" -> requires_eq_value ~label ~value (fun x -> Part x)
+  | "env" -> requires_eq_value ~label ~value (fun x -> Env x)
+  | "require-package" ->
+    requires_eq_value ~label ~value (fun x -> Require_package x)
+  | l when is_prefix ~prefix:"set-" l ->
+    requires_eq_value ~label ~value
+      (fun x -> Set (split_prefix ~prefix:"set-" l, x))
+  | l -> Error (`Msg (Format.sprintf "`%s` is not a valid label." l))
+
+let of_string s =
+  let f acc s =
+    let label, value = Relation.raw_parse s in
+    match acc, interpret label value with
+    | Ok labels, Ok label -> Ok (label :: labels)
+    | Error msgs, Ok _ -> Error msgs
+    | Ok _, Error msg -> Error [msg]
+    | Error msgs, Error msg -> Error (msg :: msgs)
+  in
+  let not_empty = function "" -> false | _ -> true in
+  let split = String.split_on_char ',' s |> List.filter not_empty in
```

Comment:
Filtering allows leading and trailing comas and double comas `",dir=a,,"`. That's not necessarily bad

Reply:
It was not allowed before, I added unit tests for that.